openpype/hosts/nuke/api/lib.py

```diff
@@ -3423,3 +3423,60 @@ def create_viewer_profile_string(viewer, display=None, path_like=False):
     if path_like:
         return "{}/{}".format(display, viewer)
     return "{} ({})".format(viewer, display)
+
+
+def get_head_filename_without_hashes(original_path, name):
+    """Function to get the renamed head filename without frame hashes
+    To avoid the system being confused on finding the filename with
+    frame hashes if the head of the filename has the hashed symbol
+
+    Args:
+        original_path (str): the filename with frame hashes
+        e.g. "renderCompositingMain.####.exr"
+        name (str): the name of the tags
+        e.g. "baking"
+
+    Returns:
+        filename: the renamed filename with the tag
+        e.g. "renderCompositingMain.baking.####.exr"
+    """
+    filename = os.path.basename(original_path)
+    fhead = os.path.splitext(filename)[0].strip(".")
+    if "#" in fhead:
+        fhead = re.sub("#+", "", fhead).rstrip(".")
+    elif "%" in fhead:
+        # use regex to convert %04d to {:0>4}
+        padding = re.search("%(\\d)+d", fhead)
+        padding = padding.group(1) if padding else 1
+        fhead = re.sub(
+            "%.*d",
+            "{{:0>{}}}".format(padding),
+            fhead
+        ).rstip(".")
```

Comment:
```suggestion
        ).rstrip(".")
```
Typo?

Reply:
nice catch

Reply:
Even with this fixed it returns:
```python
print(get_head_filename_without_hashes("renderCompositingMain.%04d.exr", "baking"))
# renderCompositingMain.%04d.exr
```
Shouldn't it include `.baking`?

Reply:
It kind of feels like what you're trying to do is this:
```python
import os
import re


def get_head_filename_without_hashes(original_path, name):
    """Function to get the renamed head filename without frame hashes
    To avoid the system being confused on finding the filename with
    frame hashes if the head of the filename has the hashed symbol

    Args:
        original_path (str): the filename with frame hashes
        e.g. "renderCompositingMain.####.exr"
        name (str): the name of the tags
        e.g. "baking"

    Returns:
        filename: the renamed filename with the tag
        e.g. "renderCompositingMain.baking.####.exr"
    """
    filename = os.path.basename(original_path)
    fhead = os.path.splitext(filename)[0].strip(".")
    
    
    def insert_name(matchobj):
        return "{}.{}".format(name, matchobj.group(0))
    
    return re.sub(r"(%\d*d)|#+", insert_name, filename)
    
    
print(get_head_filename_without_hashes("renderCompositingMain.%4d.exr", "baking"))
print(get_head_filename_without_hashes("renderCompositingMain.####.exr", "baking"))
print(get_head_filename_without_hashes("renderCompositingMain.%d.exr", "baking"))
```

Note that this would put the baking naming in front of EVERY occurence of `%4d` or `####`
```python
print(get_head_filename_without_hashes("renderCompositingMain.###.%d.exr", "baking"))
renderCompositingMain.baking.###.baking.%d.exr
```

Reply:
> Even with this fixed it returns:
> 
> ```python
> print(get_head_filename_without_hashes("renderCompositingMain.%04d.exr", "baking"))
> # renderCompositingMain.%04d.exr
> ```
> 
> Shouldn't it include `.baking`?

have you tried in Nuke?

Reply:
> Even with this fixed it returns:
> 
> ```python
> print(get_head_filename_without_hashes("renderCompositingMain.%04d.exr", "baking"))
> # renderCompositingMain.%04d.exr
> ```
> 
> Shouldn't it include `.baking`?

It should include the baking as I checked.
![image](https://github.com/ynput/OpenPype/assets/64118225/3d398dde-0c77-4d9e-9bb4-1fea8ea4b7e1)

Reply:
It doesn't work for the case with `%d` or `%04d` as per my example code.

Reply:
For` print(get_head_filename_without_hashes("renderCompositingMain.%4d.exr", "baking"))`, it is supposed it would be `renderCompositingMain.0004.baking.exr`, if I dont predict wrong.

Reply:
I also wonder whether it's worth to add pytest-able docstrings here. They are quite easy to read, make sense to a developer and can actually be run as part of your test suite.

It would look e.g. like this:

```
import os
import re


def get_head_filename_without_hashes(original_path, name):
    """Function to get the renamed head filename without frame hashes
    To avoid the system being confused on finding the filename with
    frame hashes if the head of the filename has the hashed symbol

    Examples:
        >>> get_head_filename_without_hashes("render.####.exr", "baking")
        render.baking.####.exr
        >>> get_head_filename_without_hashes("render.%d.exr", "tag")
        render.tag.%d.exr
        >>> get_head_filename_without_hashes("exr.####.exr", "foo")
        exr.foo.%04d.exr

    Args:
        original_path (str): the filename with frame hashes
        name (str): the name of the tags

    Returns:
        str: the renamed filename with the tag

    """
    filename = os.path.basename(original_path)
        
    def insert_name(matchobj):
        return "{}.{}".format(name, matchobj.group(0))
    
    return re.sub(r"(%\d*d)|#+", insert_name, filename)
```

Reply:
> > For` print(get_head_filename_without_hashes("renderCompositingMain.%4d.exr", "baking"))`, it is supposed it would be `renderCompositingMain.0004.baking.exr`, if I dont predict wrong.
> 
> Why would it be `.0004`?
> 
> Anyway - the refactored code still has the issue that the function name says it returns the "head" of the filename even though it returns the whole filename. Also, it still returns it with hashes.
> 
> All it really does is insert a particular string in front of the frame number. Still not sure why and what this is for.

it is just for example if you have frame 4 as that case....